Autoscaler fails with HTTP 403 when querying Nomad /nodes API

We are currently using the autoscaler as a Nomad job that runs inside a **namespace** and has a scaling policy that references an AWS autoscaling group for a target. The autoscaler job is using an IAM role and retrieves a Nomad ACL token from Vault. The autoscaler detects our metric and scales up the aws autoscaling group target as expected. Afterwards it fails to scale down the aws autoscaling group and we see HTTP 403 errors when it attempts to hit the /nodes API. If I take the same token the autoscaler has I can curl the /nodes API successfully. This is blocking us from fully adopting the autoscaler and feels like a namespace problem.

- nomad enterprise v1.0.2.
- autoscaler v0.2.1 

Errors:

`2021-02-11T02:47:24.458Z [ERROR] policy_eval.worker.check_handler: failed to submit scaling action to target: check=batch_job_queue_size id=c90f4d86-1485-df13-48c2-6e3bb75fd11b policy_id=5514dc5e-445d-6aae-c54c-bb629f7a1d5d queue=cluster source=datadog strategy=target-value target=aws-asg error="failed to perform scaling action: failed to perform pre-scale Nomad scale in tasks: failed to identify nodes for removal: failed to list Nomad nodes from API: Unexpected response code: 403 (Permission denied)"`

`2021-02-11T02:47:24.458Z [ERROR] policy_eval.worker: failed to evaluate policy: eval_id=4da7f9b8-91cc-3ba5-eb93-8c3634521ee4 eval_token=6c67961f-3424-c06c-a71d-7281276cec7e id=c90f4d86-1485-df13-48c2-6e3bb75fd11b policy_id=5514dc5e-445d-6aae-c54c-bb629f7a1d5d queue=cluster err="failed to scale target: failed to perform scaling action: failed to perform pre-scale Nomad scale in tasks: failed to identify nodes for removal: failed to list Nomad nodes from API: Unexpected response code: 403 (Permission denied)"`

Any help here would be appreciated!

--- a/agent/agent.go
+++ b/agent/agent.go
@@ -28,12 +28,18 @@ type Agent struct {
 	policyManager *policy.Manager
 	inMemSink     *metrics.InmemSink
 	evalBroker    *policyeval.Broker
+
+	// nomadCfg is the merged Nomad API configuration that should be used when
+	// setting up all clients. It is the result of the Nomad api.DefaultConfig
+	// merged with the user specified Nomad config.Nomad.
+	nomadCfg *api.Config
 }
 
 func NewAgent(c *config.Agent, logger hclog.Logger) *Agent {
 	return &Agent{
-		logger: logger,
-		config: c,
+		logger:   logger,
+		config:   c,
+		nomadCfg: nomadHelper.MergeDefaultWithAgentConfig(c.Nomad),
 	}
 }
 
@@ -142,55 +148,11 @@ func (a *Agent) stop() {
 	}
 }
 
-// generateNomadClient takes the internal Nomad configuration, translates and
-// merges it into a Nomad API config object and creates a client.
+// generateNomadClient creates a Nomad client for use within the agent.
 func (a *Agent) generateNomadClient() error {
 
-	// Use the Nomad API default config which gets populated by defaults and
-	// also checks for environment variables.
-	cfg := api.DefaultConfig()
-
-	// Merge our top level configuration options in.
-	if a.config.Nomad.Address != "" {
-		cfg.Address = a.config.Nomad.Address
-	}
-	if a.config.Nomad.Region != "" {
-		cfg.Region = a.config.Nomad.Region
-	}
-	if a.config.Nomad.Namespace != "" {
-		cfg.Namespace = a.config.Nomad.Namespace
-	}
-	if a.config.Nomad.Token != "" {
-		cfg.SecretID = a.config.Nomad.Token
-	}
-
-	// Merge HTTP auth.
-	if a.config.Nomad.HTTPAuth != "" {
-		cfg.HttpAuth = nomadHelper.HTTPAuthFromString(a.config.Nomad.HTTPAuth)
-	}
-
-	// Merge TLS.
-	if a.config.Nomad.CACert != "" {
-		cfg.TLSConfig.CACert = a.config.Nomad.CACert
-	}
-	if a.config.Nomad.CAPath != "" {
-		cfg.TLSConfig.CAPath = a.config.Nomad.CAPath
-	}
-	if a.config.Nomad.ClientCert != "" {
-		cfg.TLSConfig.ClientCert = a.config.Nomad.ClientCert
-	}
-	if a.config.Nomad.ClientKey != "" {
-		cfg.TLSConfig.ClientKey = a.config.Nomad.ClientKey
-	}
-	if a.config.Nomad.TLSServerName != "" {
-		cfg.TLSConfig.TLSServerName = a.config.Nomad.TLSServerName
-	}
-	if a.config.Nomad.SkipVerify {
-		cfg.TLSConfig.Insecure = a.config.Nomad.SkipVerify
-	}
-
 	// Generate the Nomad client.
-	client, err := api.NewClient(cfg)
+	client, err := api.NewClient(a.nomadCfg)
 	if err != nil {
 		return fmt.Errorf("failed to instantiate Nomad client: %v", err)
 	}

--- a/agent/agent_test.go
+++ b/agent/agent_test.go
@@ -1,0 +1,41 @@
+package agent
+
+import (
+	"errors"
+	"github.com/hashicorp/nomad/api"
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
+func TestAgent_generateNomadClient(t *testing.T) {
+	testCases := []struct {
+		inputAgent       *Agent
+		expectedOutputEr error
+		name             string
+	}{
+		{
+			inputAgent: &Agent{
+				nomadCfg: api.DefaultConfig(),
+			},
+			expectedOutputEr: nil,
+			name:             "default Nomad API config input",
+		},
+		{
+			inputAgent: &Agent{
+				nomadCfg: &api.Config{
+					Address: "\t",
+				},
+			},
+			expectedOutputEr: errors.New(`failed to instantiate Nomad client: invalid address '	': parse "\t": net/url: invalid control character in URL`),
+			name:             "invalid input Nomad address",
+		},
+	}
+
+	for _, tc := range testCases {
+		actualOutputErr := tc.inputAgent.generateNomadClient()
+		assert.Equal(t, tc.expectedOutputEr, actualOutputErr, tc.name)
+		if actualOutputErr == nil {
+			assert.Equal(t, tc.inputAgent.nomadCfg.Address, tc.inputAgent.nomadClient.Address(), tc.name)
+		}
+	}
+}

--- a/agent/plugins.go
+++ b/agent/plugins.go
@@ -62,7 +62,7 @@ func (a *Agent) setupPluginConfig(cfg map[string]string) {
 	// Nomad config from the agent. If we do not find it, opt-in by default.
 	val, ok := cfg[plugins.ConfigKeyNomadConfigInherit]
 	if !ok {
-		nomadHelper.MergeMapWithAgentConfig(cfg, a.config.Nomad)
+		nomadHelper.MergeMapWithAgentConfig(cfg, a.nomadCfg)
 		return
 	}
 
@@ -76,7 +76,7 @@ func (a *Agent) setupPluginConfig(cfg map[string]string) {
 		return
 	}
 	if boolVal {
-		nomadHelper.MergeMapWithAgentConfig(cfg, a.config.Nomad)
+		nomadHelper.MergeMapWithAgentConfig(cfg, a.nomadCfg)
 	}
 }
 

--- a/agent/plugins_test.go
+++ b/agent/plugins_test.go
@@ -3,8 +3,9 @@ package agent
 import (
 	"testing"
 
-	"github.com/hashicorp/go-hclog"
+	hclog "github.com/hashicorp/go-hclog"
 	"github.com/hashicorp/nomad-autoscaler/agent/config"
+	"github.com/hashicorp/nomad/api"
 	"github.com/stretchr/testify/assert"
 )
 
@@ -21,19 +22,22 @@ func TestAgent_setupPluginConfig(t *testing.T) {
 			},
 			inputAgent: &Agent{
 				logger: hclog.NewNullLogger(),
-				config: &config.Agent{
-					Nomad: &config.Nomad{
-						Address:       "test",
-						Region:        "test",
-						Namespace:     "test",
-						Token:         "test",
-						HTTPAuth:      "test",
+				nomadCfg: &api.Config{
+					Address:   "test",
+					Region:    "test",
+					SecretID:  "test",
+					Namespace: "test",
+					HttpAuth: &api.HttpBasicAuth{
+						Username: "test",
+						Password: "test",
+					},
+					TLSConfig: &api.TLSConfig{
 						CACert:        "test",
 						CAPath:        "test",
 						ClientCert:    "test",
 						ClientKey:     "test",
 						TLSServerName: "test",
-						SkipVerify:    true,
+						Insecure:      true,
 					},
 				},
 			},
@@ -48,19 +52,22 @@ func TestAgent_setupPluginConfig(t *testing.T) {
 			},
 			inputAgent: &Agent{
 				logger: hclog.NewNullLogger(),
-				config: &config.Agent{
-					Nomad: &config.Nomad{
-						Address:       "test",
-						Region:        "test",
-						Namespace:     "test",
-						Token:         "test",
-						HTTPAuth:      "test",
+				nomadCfg: &api.Config{
+					Address:   "test",
+					Region:    "test",
+					SecretID:  "test",
+					Namespace: "test",
+					HttpAuth: &api.HttpBasicAuth{
+						Username: "test",
+						Password: "test",
+					},
+					TLSConfig: &api.TLSConfig{
 						CACert:        "test",
 						CAPath:        "test",
 						ClientCert:    "test",
 						ClientKey:     "test",
 						TLSServerName: "test",
-						SkipVerify:    true,
+						Insecure:      true,
 					},
 				},
 			},
@@ -75,19 +82,22 @@ func TestAgent_setupPluginConfig(t *testing.T) {
 			},
 			inputAgent: &Agent{
 				logger: hclog.NewNullLogger(),
-				config: &config.Agent{
-					Nomad: &config.Nomad{
-						Address:       "test",
-						Region:        "test",
-						Namespace:     "test",
-						Token:         "test",
-						HTTPAuth:      "test",
+				nomadCfg: &api.Config{
+					Address:   "test",
+					Region:    "test",
+					SecretID:  "test",
+					Namespace: "test",
+					HttpAuth: &api.HttpBasicAuth{
+						Username: "test",
+						Password: "test",
+					},
+					TLSConfig: &api.TLSConfig{
 						CACert:        "test",
 						CAPath:        "test",
 						ClientCert:    "test",
 						ClientKey:     "test",
 						TLSServerName: "test",
-						SkipVerify:    true,
+						Insecure:      true,
 					},
 				},
 			},
@@ -97,7 +107,7 @@ func TestAgent_setupPluginConfig(t *testing.T) {
 				"nomad_region":          "test",
 				"nomad_namespace":       "test",
 				"nomad_token":           "test",
-				"nomad_http-auth":       "test",
+				"nomad_http-auth":       "test:test",
 				"nomad_ca-cert":         "test",
 				"nomad_ca-path":         "test",
 				"nomad_client-cert":     "test",
@@ -111,19 +121,22 @@ func TestAgent_setupPluginConfig(t *testing.T) {
 			inputCfg: map[string]string{},
 			inputAgent: &Agent{
 				logger: hclog.NewNullLogger(),
-				config: &config.Agent{
-					Nomad: &config.Nomad{
-						Address:       "test",
-						Region:        "test",
-						Namespace:     "test",
-						Token:         "test",
-						HTTPAuth:      "test",
+				nomadCfg: &api.Config{
+					Address:   "test",
+					Region:    "test",
+					SecretID:  "test",
+					Namespace: "test",
+					HttpAuth: &api.HttpBasicAuth{
+						Username: "test",
+						Password: "test",
+					},
+					TLSConfig: &api.TLSConfig{
 						CACert:        "test",
 						CAPath:        "test",
 						ClientCert:    "test",
 						ClientKey:     "test",
 						TLSServerName: "test",
-						SkipVerify:    true,
+						Insecure:      true,
 					},
 				},
 			},
@@ -132,7 +145,7 @@ func TestAgent_setupPluginConfig(t *testing.T) {
 				"nomad_region":          "test",
 				"nomad_namespace":       "test",
 				"nomad_token":           "test",
-				"nomad_http-auth":       "test",
+				"nomad_http-auth":       "test:test",
 				"nomad_ca-cert":         "test",
 				"nomad_ca-path":         "test",
 				"nomad_client-cert":     "test",

--- a/sdk/helper/nomad/config.go
+++ b/sdk/helper/nomad/config.go
@@ -92,44 +92,104 @@ func HTTPAuthFromString(auth string) *api.HttpBasicAuth {
 	}
 }
 
-// MergeMapWithAgentConfig merges a Nomad map config with an agent config. The
-// config parameters within the map take preference.
-func MergeMapWithAgentConfig(m map[string]string, agentCfg *config.Nomad) {
-	if agentCfg == nil {
+// MergeMapWithAgentConfig merges a Nomad map config with an API config object
+// with the map config taking precedence. This allows users to override only a
+// subset of params, while inheriting the agent configured items which are also
+// derived from Nomad API default and env vars.
+func MergeMapWithAgentConfig(m map[string]string, cfg *api.Config) {
+	if cfg == nil {
 		return
 	}
 
-	if agentCfg.Address != "" && m[configKeyNomadAddress] == "" {
-		m[configKeyNomadAddress] = agentCfg.Address
+	if cfg.Address != "" && m[configKeyNomadAddress] == "" {
+		m[configKeyNomadAddress] = cfg.Address
 	}
-	if agentCfg.Region != "" && m[configKeyNomadRegion] == "" {
-		m[configKeyNomadRegion] = agentCfg.Region
+	if cfg.Region != "" && m[configKeyNomadRegion] == "" {
+		m[configKeyNomadRegion] = cfg.Region
 	}
-	if agentCfg.Namespace != "" && m[configKeyNomadNamespace] == "" {
-		m[configKeyNomadNamespace] = agentCfg.Namespace
+	if cfg.Namespace != "" && m[configKeyNomadNamespace] == "" {
+		m[configKeyNomadNamespace] = cfg.Namespace
 	}
-	if agentCfg.Token != "" && m[configKeyNomadToken] == "" {
-		m[configKeyNomadToken] = agentCfg.Token
+	if cfg.SecretID != "" && m[configKeyNomadToken] == "" {
+		m[configKeyNomadToken] = cfg.SecretID
 	}
-	if agentCfg.CACert != "" && m[configKeyNomadCACert] == "" {
-		m[configKeyNomadCACert] = agentCfg.CACert
+	if cfg.TLSConfig.CACert != "" && m[configKeyNomadCACert] == "" {
+		m[configKeyNomadCACert] = cfg.TLSConfig.CACert
 	}
-	if agentCfg.CAPath != "" && m[configKeyNomadCAPath] == "" {
-		m[configKeyNomadCAPath] = agentCfg.CAPath
+	if cfg.TLSConfig.CAPath != "" && m[configKeyNomadCAPath] == "" {
+		m[configKeyNomadCAPath] = cfg.TLSConfig.CAPath
 	}
-	if agentCfg.ClientCert != "" && m[configKeyNomadClientCert] == "" {
-		m[configKeyNomadClientCert] = agentCfg.ClientCert
+	if cfg.TLSConfig.ClientCert != "" && m[configKeyNomadClientCert] == "" {
+		m[configKeyNomadClientCert] = cfg.TLSConfig.ClientCert
 	}
-	if agentCfg.ClientKey != "" && m[configKeyNomadClientKey] == "" {
-		m[configKeyNomadClientKey] = agentCfg.ClientKey
+	if cfg.TLSConfig.ClientKey != "" && m[configKeyNomadClientKey] == "" {
+		m[configKeyNomadClientKey] = cfg.TLSConfig.ClientKey
 	}
-	if agentCfg.TLSServerName != "" && m[configKeyNomadTLSServerName] == "" {
-		m[configKeyNomadTLSServerName] = agentCfg.TLSServerName
+	if cfg.TLSConfig.TLSServerName != "" && m[configKeyNomadTLSServerName] == "" {
+		m[configKeyNomadTLSServerName] = cfg.TLSConfig.TLSServerName
 	}
-	if agentCfg.SkipVerify && m[configKeyNomadSkipVerify] == "" {
-		m[configKeyNomadSkipVerify] = strconv.FormatBool(agentCfg.SkipVerify)
+	if cfg.TLSConfig.Insecure && m[configKeyNomadSkipVerify] == "" {
+		m[configKeyNomadSkipVerify] = strconv.FormatBool(cfg.TLSConfig.Insecure)
 	}
-	if agentCfg.HTTPAuth != "" && m[configKeyNomadHTTPAuth] == "" {
-		m[configKeyNomadHTTPAuth] = agentCfg.HTTPAuth
+	if cfg.HttpAuth != nil && m[configKeyNomadHTTPAuth] == "" {
+		auth := cfg.HttpAuth.Username
+		if cfg.HttpAuth.Password != "" {
+			auth += ":" + cfg.HttpAuth.Password
+		}
+		m[configKeyNomadHTTPAuth] = auth
 	}
+}
+
+// MergeDefaultWithAgentConfig merges the agent Nomad configuration with the
+// default Nomad API configuration. The Nomad Autoscaler agent config takes
+// precedence over the default config as any user supplied variables should
+// override those configured by default or discovered via env vars within the
+// Nomad API config.
+func MergeDefaultWithAgentConfig(agentCfg *config.Nomad) *api.Config {
+
+	// Use the Nomad API default config which gets populated by defaults and
+	// also checks for environment variables.
+	cfg := api.DefaultConfig()
+
+	// Merge our top level configuration options in.
+	if agentCfg.Address != "" {
+		cfg.Address = agentCfg.Address
+	}
+	if agentCfg.Region != "" {
+		cfg.Region = agentCfg.Region
+	}
+	if agentCfg.Namespace != "" {
+		cfg.Namespace = agentCfg.Namespace
+	}
+	if agentCfg.Token != "" {
+		cfg.SecretID = agentCfg.Token
+	}
+
+	// Merge HTTP auth.
+	if agentCfg.HTTPAuth != "" {
+		cfg.HttpAuth = HTTPAuthFromString(agentCfg.HTTPAuth)
+	}
+
+	// Merge TLS. The default config has an empty TLS object and therefore does
+	// not required a nil check.
+	if agentCfg.CACert != "" {
+		cfg.TLSConfig.CACert = agentCfg.CACert
+	}
+	if agentCfg.CAPath != "" {
+		cfg.TLSConfig.CAPath = agentCfg.CAPath
+	}
+	if agentCfg.ClientCert != "" {
+		cfg.TLSConfig.ClientCert = agentCfg.ClientCert
+	}
+	if agentCfg.ClientKey != "" {
+		cfg.TLSConfig.ClientKey = agentCfg.ClientKey
+	}
+	if agentCfg.TLSServerName != "" {
+		cfg.TLSConfig.TLSServerName = agentCfg.TLSServerName
+	}
+	if agentCfg.SkipVerify {
+		cfg.TLSConfig.Insecure = agentCfg.SkipVerify
+	}
+
+	return cfg
 }

--- a/sdk/helper/nomad/config_test.go
+++ b/sdk/helper/nomad/config_test.go
@@ -82,31 +82,36 @@ func Test_HTTPAuthFromString(t *testing.T) {
 func Test_MergeMapWithAgentConfig(t *testing.T) {
 	testCases := []struct {
 		inputMap          map[string]string
-		inputAgentConfig  *config.Nomad
+		inputAPIConfig    *api.Config
 		expectedOutputMap map[string]string
 		name              string
 	}{
 		{
 			inputMap: map[string]string{},
-			inputAgentConfig: &config.Nomad{
-				Address:       "test",
-				Region:        "test",
-				Namespace:     "test",
-				Token:         "test",
-				HTTPAuth:      "test",
-				CACert:        "test",
-				CAPath:        "test",
-				ClientCert:    "test",
-				ClientKey:     "test",
-				TLSServerName: "test",
-				SkipVerify:    true,
+			inputAPIConfig: &api.Config{
+				Address:   "test",
+				Region:    "test",
+				Namespace: "test",
+				SecretID:  "test",
+				HttpAuth: &api.HttpBasicAuth{
+					Username: "test",
+					Password: "test",
+				},
+				TLSConfig: &api.TLSConfig{
+					CACert:        "test",
+					CAPath:        "test",
+					ClientCert:    "test",
+					ClientKey:     "test",
+					TLSServerName: "test",
+					Insecure:      true,
+				},
 			},
 			expectedOutputMap: map[string]string{
 				"nomad_address":         "test",
 				"nomad_region":          "test",
 				"nomad_namespace":       "test",
 				"nomad_token":           "test",
-				"nomad_http-auth":       "test",
+				"nomad_http-auth":       "test:test",
 				"nomad_ca-cert":         "test",
 				"nomad_ca-path":         "test",
 				"nomad_client-cert":     "test",
@@ -116,55 +121,67 @@ func Test_MergeMapWithAgentConfig(t *testing.T) {
 			},
 			name: "empty input map",
 		},
+	}
 
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			MergeMapWithAgentConfig(tc.inputMap, tc.inputAPIConfig)
+			assert.Equal(t, tc.expectedOutputMap, tc.inputMap, tc.name)
+		})
+	}
+}
+
+func Test_MergeDefaultWithAgentConfig(t *testing.T) {
+	testCases := []struct {
+		inputConfig    *config.Nomad
+		expectedOutput *api.Config
+		name           string
+	}{
 		{
-			inputMap: map[string]string{
-				"nomad_address":         "test",
-				"nomad_region":          "test",
-				"nomad_namespace":       "test",
-				"nomad_token":           "test",
-				"nomad_http-auth":       "test",
-				"nomad_ca-cert":         "test",
-				"nomad_ca-path":         "test",
-				"nomad_client-cert":     "test",
-				"nomad_client-key":      "test",
-				"nomad_tls-server-name": "test",
-				"nomad_skip-verify":     "true",
+			inputConfig:    &config.Nomad{},
+			expectedOutput: api.DefaultConfig(),
+			name:           "default Autoscaler Nomad config",
+		},
+		{
+			inputConfig: &config.Nomad{
+				Address:       "http://demo.nomad:4646",
+				Region:        "vlc",
+				Namespace:     "platform",
+				Token:         "shhhhhhhh",
+				HTTPAuth:      "admin:admin",
+				CACert:        "/path/to/long-lived/ca-cert",
+				CAPath:        "/path/to/long-lived/",
+				ClientCert:    "/path/to/long-lived/client-cert",
+				ClientKey:     "/path/to/long-lived/key-cert",
+				TLSServerName: "whatdoesthisdo",
+				SkipVerify:    true,
 			},
-			inputAgentConfig: &config.Nomad{
-				Address:       "test-new",
-				Region:        "test-new",
-				Namespace:     "test-new",
-				Token:         "test-new",
-				HTTPAuth:      "test-new",
-				CACert:        "test-new",
-				CAPath:        "test-new",
-				ClientCert:    "test-new",
-				ClientKey:     "test-new",
-				TLSServerName: "test-new",
-				SkipVerify:    false,
+			expectedOutput: &api.Config{
+				Address:   "http://demo.nomad:4646",
+				Region:    "vlc",
+				SecretID:  "shhhhhhhh",
+				Namespace: "platform",
+				HttpAuth: &api.HttpBasicAuth{
+					Username: "admin",
+					Password: "admin",
+				},
+				TLSConfig: &api.TLSConfig{
+					CACert:        "/path/to/long-lived/ca-cert",
+					CAPath:        "/path/to/long-lived/",
+					ClientCert:    "/path/to/long-lived/client-cert",
+					ClientKey:     "/path/to/long-lived/key-cert",
+					TLSServerName: "whatdoesthisdo",
+					Insecure:      true,
+				},
 			},
-			expectedOutputMap: map[string]string{
-				"nomad_address":         "test",
-				"nomad_region":          "test",
-				"nomad_namespace":       "test",
-				"nomad_token":           "test",
-				"nomad_http-auth":       "test",
-				"nomad_ca-cert":         "test",
-				"nomad_ca-path":         "test",
-				"nomad_client-cert":     "test",
-				"nomad_client-key":      "test",
-				"nomad_tls-server-name": "test",
-				"nomad_skip-verify":     "true",
-			},
-			name: "fully populated input map and input agent config",
+			name: "full Autoscaler Nomad config override",
 		},
 	}
 
 	for _, tc := range testCases {
 		t.Run(tc.name, func(t *testing.T) {
-			MergeMapWithAgentConfig(tc.inputMap, tc.inputAgentConfig)
-			assert.Equal(t, tc.expectedOutputMap, tc.inputMap, tc.name)
+			actualOutput := MergeDefaultWithAgentConfig(tc.inputConfig)
+			assert.Equal(t, tc.expectedOutput, actualOutput, tc.name)
 		})
 	}
 }

